### DISCUSSION TITLE
replacing def_buffer

### DISCUSSION THREAD

**Discussion by fraguada** [wjakob/nanobind#699] [Q&A]
I'm working on switching our lib to use nanobind and don't have a good idea of how to replace def_buffer. Here is an example of what we have running in pybind:

From: https://github.com/mcneel/rhino3dm/blob/main/src/bindings/bnd_nurbscurve.cpp#L255
```cpp
py::class_<BND_NurbsCurvePointList>(m, "NurbsCurvePointList", py::buffer_protocol())
    // ...
    .def_buffer([](BND_NurbsCurvePointList& pl) -> py::buffer_info
    {
      return py::buffer_info
      (
        pl.GetCurve()->m_cv,                      /* Pointer to buffer */
        sizeof(double),                           /* Size of one scalar */
        py::format_descriptor<double>::format(),  /* Python struct-style format descriptor */
        2,                                        /* Number of dimensions */
        {pl.Count(), pl.GetCVDims() },            /* Buffer dimensions */
        {pl.GetCurve()->m_cv_stride * sizeof(double), sizeof(double)}  /* Strides (in bytes) for each index */
      );
    })
    //...
    ;
```
Would appreciate any tips on how to do this in nanobind!

**Comment by wojdyr** [comment]
FTR a similar question was asked by @awni in #478

**Comment by awni** [comment]
We did this in MLX. Here's the relevant code if it's useful:

- Insert the right functions in the object slots https://github.com/ml-explore/mlx/blob/main/python/src/array.cpp#L209-L213
- Implement the buffer protocol functions https://github.com/ml-explore/mlx/blob/main/python/src/buffer.h

**Reply by fraguada** [reply]
Thanks for the tip. We'll study how to implement the buffer protocol functions in our code base based on your implementation and report back when we have something working.

**Comment by soggy-cr0uton** [comment]
Closing as self-resolved — the behavior is expected due to post-completion auto cross-seed search, not the exclude categories setting.
